File: math/16.3SumClosest.py
import sys
from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums = sorted(nums)
        res = 0
        data = sys.maxsize
        for i, num in enumerate(nums):
            l = i + 1
            r = len(nums) - 1
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            while l < r:
                _sum = num + nums[l] + nums[r]
                if target == _sum:
                    return target
                if target < _sum:
                    r -= 1
                    if data > _sum - target:
                        data = _sum - target
                        res = _sum
                else:
                    l += 1
                    if data > target - _sum:
                        data = target - _sum
                        res = _sum
        return res


if __name__ == "__main__":
    s = Solution()

    nums = [-1, 2, 1, -4]
    target = 1
    # target is 2. (-1 + 2 + 1 = 2).
    print(s.threeSumClosest(nums, target))

    nums = [1, 1, 1, 1]
    target = 0
    print(s.threeSumClosest(nums, target))

File: math/16.3SumClosest.go
package main

import (
	"math"
	"sort"
)

func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	res, data := 0, math.MaxInt64

	for i := range nums {
		j, k := i+1, len(nums)-1
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		for j < k {
			sum := nums[i] + nums[j] + nums[k]
			if sum == target {
				return sum
			}
			if sum > target {
				k--
				if data > sum-target {
					data = sum - target
					res = sum
				}
			} else if sum < target {
				j++
				if data > target-sum {
					data = target - sum
					res = sum
				}
			}
		}
	}
	return res
}
